### PR TITLE
Document how to set up UPnP in a non-blocking manner

### PR DESCRIPTION
**Pull request by Calinou**
See https://github.com/godotengine/godot/issues/32160. Thanks @Faless for providing the code sample which I modified a bit :slightly_smiling_face:

### CODE REVIEW

File: modules/upnp/doc_classes/UPNP.xml [hunk 1]
```diff
@@ -16,6 +16,32 @@
 		[codeblock]
 		upnp.delete_port_mapping(port)
 		[/codeblock]
+		[b]Note:[/b] UPnP discovery blocks the main thread. To perform non-blocking discovery, use [Thread]s like this:
```
Comment (Faless) [inline modules/upnp/doc_classes/UPNP.xml:19]: > UPnP discovery blocks the main thread.

"UPnP discovery blocks the current thread. To perform discovery without blocking the main thread use [Thread]s like this:"

File: modules/upnp/doc_classes/UPNP.xml [hunk 2]
```diff
@@ -16,6 +16,32 @@
...
+		const SERVER_PORT = 3928
+		var thread = null
+
+		func _upnp_setup(server_port):
+		    # UPNP queries take some time.
+		    var upnp = UPNP.new()
+		    var err = upnp.discover()
+
+		    if err:
+		        return push_error(err)
```
Comment (Faless) [inline modules/upnp/doc_classes/UPNP.xml:31]: `push_error` returns void.
`push_error` also seems to fail (causing the script to crash) when passed a non-string argument.

```gdscript
	if err != OK:
		push_error(str(err))
		return
```

**Review by Faless — DISMISSED** [review]
See comments, we might also add a variable to check when the mapping is done.
It's hard to see the limit to what should be made as an example and what not.
E.g. I didn't feel this needed a specific example, as long as there is a generic example of using threads, because this just how you normally use them.

**Comment by Calinou** [comment]
> See comments, we might also add a variable to check when the mapping is done.

Should this be a signal emitted instead of a `upnp_mapping_set` variable set to `true`?

**Comment by Faless** [comment]
> Should this be a signal emitted instead of a `upnp_mapping_set` variable being set to `true`?

Yes, that's a good idea. Maybe a `upnp_completed` signal with an `Error` parameter (`OK` when done successfully or any error in case of failure, like done in HTTPRequest)?

**Comment by mhilbrunner** [comment]
@Calinou Friendly reminder that this needs a small update :)

**Comment by Calinou** [comment]
I added the `upnp_completed` signal emission as requested :slightly_smiling_face:

**Review by mhilbrunner — APPROVED** [review]

**Comment by mhilbrunner** [comment]
Thanks for this!

**Comment by akien-mga** [comment]
Cherry-picked for 3.4.

**Pull request merged**